File: 139-word-break/139-word-break.java
class Solution {
    public boolean wordBreak(String s, List<String> wordDict) {
        HashMap<String,Boolean> hm = new HashMap<>();
        return helper(s,wordDict,hm);
    }
    
    public boolean helper(String s, List<String> wordDict,HashMap<String,Boolean> hm) {
        if(hm.containsKey(s)) return hm.get(s);
        if(s=="")return true;
        
        for(String word:wordDict){
            if(s.indexOf(word)==0){
                String suffix = s.substring(word.length());
                Boolean result = helper(suffix, wordDict,hm);
                hm.put(suffix,result);
                if(result) return true;
            }
        }
        hm.put(s,false);
        return false;
    }
}

File: 139-word-break/139-word-break.js
/**
 * @param {string} s
 * @param {string[]} wordDict
 * @return {boolean}
 */
var wordBreak = function(s, wordDict) {
    const dp = [];
    return helper(s, wordDict, dp);
};

function helper(str, dict, dp){
    if(str==="") return true;
    if(dp[str.length]===1) return true;
    if(dp[str.length]===-1) return false;
    for(let word of dict){
        let idx = str.indexOf(word);
        str.substring(idx, word.length)
        if(idx === 0){
            let result = helper(str.substring(idx+word.length), dict, dp);
            dp[str.length] = result===true?1:-1;
            if(result) return true;
        }
    }
    dp[str.length] = -1;
    return false;
}